File: src/lib/code/mcs-216/section-1/session-5/1/1.c
#include <stdio.h>

/**
 * Recursive Binary Search
 * @param arr    Sorted integer array
 * @param low    Starting index of current search range
 * @param high   Ending index of current search range
 * @param target Value to search for
 * @return       Index of target, or -1 if not found
 */
int binarySearch(int arr[], int low, int high, int target) {
    // Base case: invalid range means target is not in array
    if (low > high) {
        return -1;
    }

    // Calculate mid to avoid integer overflow
    int mid = low + (high - low) / 2;

    // Print step-by-step trace for lab documentation
    printf("[Call] low=%d, high=%d, mid=%d, arr[mid]=%d\n", low, high, mid, arr[mid]);

    // Check if target is present at mid
    if (arr[mid] == target) {
        return mid;
    }

    // If target is greater, ignore left half and search right
    if (arr[mid] < target) {
        return binarySearch(arr, mid + 1, high, target);
    }

    // If target is smaller, ignore right half and search left
    return binarySearch(arr, low, mid - 1, target);
}

int main() {
    int arr[] = {10, 35, 40, 45, 50, 55, 60, 65, 70, 100};
    int n = sizeof(arr) / sizeof(arr[0]);
    int target = 100;

    printf("=== Recursive Binary Search in C ===\n");
    int result = binarySearch(arr, 0, n - 1, target);

    if (result != -1)
        printf("\n✅ Target %d found at index %d\n", target, result);
    else
        printf("\n❌ Target %d not found in array\n", target);

    return 0;
}

File: src/lib/code/mcs-216/section-1/session-5/1/1.py
def binary_search(arr, low, high, target):
    """
    Recursive Binary Search in Python
    :param arr: Sorted list of integers
    :param low: Starting index
    :param high: Ending index
    :param target: Value to find
    :return: Index of target, or -1 if not found
    """
    # Base case: search range is invalid
    if low > high:
        return -1

    # Integer division for mid calculation
    mid = low + (high - low) // 2

    # Step-by-step trace for lab submission
    print(f"[Call] low={low}, high={high}, mid={mid}, arr[mid]={arr[mid]}")

    # Base case: target found
    if arr[mid] == target:
        return mid
    # Target is in the right half
    elif arr[mid] < target:
        return binary_search(arr, mid + 1, high, target)
    # Target is in the left half
    else:
        return binary_search(arr, low, mid - 1, target)

# Driver code
if __name__ == "__main__":
    arr = [10, 35, 40, 45, 50, 55, 60, 65, 70, 100]
    target = 100
    print("=== Recursive Binary Search in Python ===")

    result = binary_search(arr, 0, len(arr) - 1, target)

    if result != -1:
        print(f"\n✅ Target {target} found at index {result}")
    else:
        print(f"\n❌ Target {target} not found in array")
